jsm.go

```diff
@@ -701,30 +706,63 @@ func (js *js) StreamInfo(stream string, opts ...JSOpt) (*StreamInfo, error) {
 	if cancel != nil {
 		defer cancel()
 	}
+
+	var i int
+	var subjectMessagesMap map[string]uint64 = nil
 	var req []byte
+
+	siOpts := StreamInfoRequest{}
 	if o.streamInfoOpts != nil {
-		if req, err = json.Marshal(o.streamInfoOpts); err != nil {
+		siOpts.DeletedDetails = o.streamInfoOpts.DeletedDetails
+		siOpts.SubjectsFilter = o.streamInfoOpts.SubjectsFilter
+	}
+
+	for {
+		siOpts.Offset = i
+		if req, err = json.Marshal(&siOpts); err != nil {
 			return nil, err
 		}
-	}
-	siSubj := js.apiSubj(fmt.Sprintf(apiStreamInfoT, stream))
 
-	r, err := js.apiRequestWithContext(o.ctx, siSubj, req)
-	if err != nil {
-		return nil, err
-	}
-	var resp streamInfoResponse
-	if err := json.Unmarshal(r.Data, &resp); err != nil {
-		return nil, err
-	}
-	if resp.Error != nil {
-		if errors.Is(resp.Error, ErrStreamNotFound) {
-			return nil, ErrStreamNotFound
+		siSubj := js.apiSubj(fmt.Sprintf(apiStreamInfoT, stream))
+
+		r, err := js.apiRequestWithContext(o.ctx, siSubj, req)
+		if err != nil {
+			return nil, err
 		}
-		return nil, resp.Error
-	}
 
-	return resp.StreamInfo, nil
+		var resp streamInfoResponse
+
+		if err := json.Unmarshal(r.Data, &resp); err != nil {
+			return nil, err
+		}
+		if resp.Error != nil {
+			if errors.Is(resp.Error, ErrStreamNotFound) {
+				return nil, ErrStreamNotFound
+			}
+			return nil, resp.Error
+		}
+
+		// for backwards compatibility
+		var total int
+		if resp.Total != 0 {
+			total = resp.Total
+		} else {
+			total = len(resp.State.Subjects)
+		}
+
+		if subjectMessagesMap == nil {
+			subjectMessagesMap = make(map[string]uint64, total)
```

Comment:
Technically, here even if user did not ask for details, you are now creating this map and returning as a resp.StreamInfo.State.Subjects (empty, but non nil map). More reasons maybe to treat the two situations differently, or at the very least do the creation only if map is nil AND the aforementioned boolean is set (indicating that we were asking for subject details)